libbeat/autodiscover/provider.go

```diff
@@ -0,0 +1,88 @@
+package autodiscover
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/cfgfile"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Provider for autodiscovery
+type Provider interface {
+	cfgfile.Runner
+}
+
+// Registry holds all known autodiscover providers, they must be added to it to enable them for use
+var Registry = NewRegistry()
+
+// ProviderBuilder creates a new provider based on the given config and returns it
+type ProviderBuilder func(bus.Bus, *common.Config) (Provider, error)
+
+// Register of autodiscovery providers
+type registry struct {
+	// Lock to control concurrent read/writes
+	lock sync.RWMutex
+	// A map of provider name to ProviderBuilder.
+	providers map[string]ProviderBuilder
+}
+
+// NewRegistry creates and returns a new Registry
+func NewRegistry() *registry {
+	return &registry{
+		providers: make(map[string]ProviderBuilder, 0),
+	}
+}
+
+// AddProvider registers a new ProviderBuilder
+func (r *registry) AddProvider(name string, provider ProviderBuilder) error {
+	r.lock.Lock()
```

Comment:
Why do you lock here? Are `init` functions called concurrently?

Reply:
Nope, or not at the moment :innocent:, but some defensive code there won't hurt I think, this may be a thing in the future?